sdcflows/interfaces/bspline.py

```diff
@@ -264,24 +266,28 @@ def _run_interface(self, runtime):
         )
 
         # Generate warp field
-        phaseEncDim = "ijk".index(self.inputs.pe_dir[0])
-        phaseEncSign = [1.0, -1.0][len(self.inputs.pe_dir) != 2]
-
-        data *= phaseEncSign * self.inputs.ro_time
+        pe_axis = "ijk".index(self.inputs.pe_dir[0])
+        pe_sign = -1.0 if self.inputs.pe_dir.endswith("-") else 1.0
+        pe_size = targetnii.header.get_zooms()[pe_axis]
+        data *= pe_sign * self.inputs.ro_time * pe_size
 
         fieldshape = tuple(list(data.shape[:3]) + [3])
         self._results["out_warp"] = fname_presuffix(
             self.inputs.in_target, suffix="_xfm", newpath=runtime.cwd
         )
         # Compose a vector field
         field = np.zeros((data.size, 3), dtype="float32")
-        field[..., phaseEncDim] = data.reshape(-1)
-        aff = targetnii.affine.copy()
-        aff[:3, 3] = 0.0
-        # Multiplying by the affine implicitly applies the voxel size to the shift map
-        field = nb.affines.apply_affine(aff, field).reshape(fieldshape)
+        field[..., pe_axis] = data.reshape(-1)
+
+        # If coordinate system is oblique, project displacements through directions matrix
+        aff = targetnii.affine
+        if (obliquity(aff).min() * 180 / pi) > 0.01:
+            dirmat = np.eye(4)
+            dirmat[:3, :3] = aff[:3, :3] / (voxel_sizes(aff) * io_orientation(aff)[:, 1])
+            field = nb.affines.apply_affine(aff, field)
```

Comment:
I'm confused. `dirmat` isn't used. Should this be:

```suggestion
            dirmat = np.eye(4)
            dirmat[:3, :3] = aff[:3, :3] / (voxel_sizes(aff) * io_orientation(aff)[:, 1])
            field = nb.affines.apply_affine(dirmat, field)
```

I'm not quite following `dirmat`. We're getting the cosines, divided by the zooms, so it should be a unit rotation matrix, and then flipping R/L, A/P, S/I. Is this flip going to undo the direction encoded in `j`/`j-`?

Reply:
This is exactly the problem we are fixing. If the PE direction is `j-` and the affine has `RPI` orientation, then this will drop the flip of the two last axes in this case.

Similarly, if the PE direction is `j-` and the orientation is `PLS`, the directions matrix will be made `ARS`.

In other words, the nonzero component of the fieldmap needs to be projected through the oblique matrix so that the displacements are actually done along the PE axis by tools like `antsApplyTransforms` that work on world coordinates.

Did I manage to explain this?

Reply:
And yes, not using `dirmat` was an oversight. The idea is that dirmat gives you the unit vectors of the rotated axes, right-hand coordinates (i.e., no flips).